model-engine/model_engine_server/inference/forwarding/http_forwarder.py

```diff
@@ -111,43 +113,133 @@ async def event_generator():
+            # because FastAPI will match the first route that matches the request path
+            async def predict_or_stream(
+                request: EndpointPredictV1Request,
+                background_tasks: BackgroundTasks,
+                sync_forwarder=Depends(lambda: sync_forwarders.get(route)),
+                stream_forwarder=Depends(lambda: stream_forwarders.get(route)),
+                limiter=Depends(get_concurrency_limiter),
+            ):
+                if not request.args:
+                    raise Exception("Request has no args")
+                if request.args.root.get("stream", False) and stream_forwarder:
+                    return await stream(request, stream_forwarder, limiter)
+                elif request.args.root.get("stream") is not True and sync_forwarder:
+                    return predict(request, background_tasks, sync_forwarder, limiter)
+                else:
+                    raise Exception("No forwarder configured for this route")
+
+            logger.info(f"Adding route {route}")
+            app.add_api_route(
+                path=route,
+                endpoint=predict_or_stream,
+                methods=["POST"],
+            )
+
+    app.add_api_route(path="/healthz", endpoint=healthcheck, methods=["GET"])
+    app.add_api_route(path="/readyz", endpoint=healthcheck, methods=["GET"])
+    app.add_api_route(path="/predict", endpoint=predict, methods=["POST"])
+    app.add_api_route(path="/stream", endpoint=stream, methods=["POST"])
+
+    add_extra_routes(app)
+    return app
+
+
 def entrypoint():
     parser = argparse.ArgumentParser()
     parser.add_argument("--config", type=str, required=True)
     parser.add_argument("--num-workers", type=int, required=True)
-    parser.add_argument("--host", type=str, default="[::]")
+    parser.add_argument("--host", type=str, default=None)
     parser.add_argument("--port", type=int, default=5000)
     parser.add_argument("--set", type=str, action="append")
     parser.add_argument("--graceful-timeout", type=int, default=600)
 
     args, extra_args = parser.parse_known_args()
 
-    values = [f"CONFIG_FILE={args.config}"]
+    os.environ["CONFIG_FILE"] = args.config
     if args.set is not None:
-        values.append(f"CONFIG_OVERRIDES={';'.join(args.set)}")
-    envs = []
-    for v in values:
-        envs.extend(["--env", v])
-
-    command = [
-        "gunicorn",
-        "--bind",
-        f"{args.host}:{args.port}",
-        "--timeout",
-        "1200",
-        "--keep-alive",
-        "2",
-        "--worker-class",
-        "uvicorn.workers.UvicornWorker",
-        "--workers",
-        str(args.num_workers),
-        "--graceful-timeout",
-        str(args.graceful_timeout),
-        *envs,
-        "model_engine_server.inference.forwarding.http_forwarder:app",
-        *extra_args,
-    ]
-    subprocess.run(command)
+        os.environ["CONFIG_OVERRIDES"] = ";".join(args.set)
+
+    asyncio.run(
+        run_server(
```

Comment:
are we getting rid of gunicorn here? probably fine, but want to make sure this doesn't kill our ability to handle lots of concurrent requests or anything like that

Reply:
yeah i'll do some rough tests to make sure this is ok

Reply:
thanks for doing this!